apiextensions apiserver: update storage version for custom resources

roycaihw opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Implements the section "CRDs" in [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md#crds).

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add a feature in the API server to update storage version for custom resources. Enabling internal.apiserver.k8s.io/v1alpha1 API, StorageVersionAPI feature gate and APIServerIdentity feature gate are required to use this feature. 
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md
```

/sig api-machinery
/assign @sttts @caesarxuchao

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -919,6 +928,51 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
+					return nil
+				}
+			}
+			sv.OwnerReferences = append(sv.OwnerReferences, ref)
```
Append an owner reference to delete stale StorageVersions. This does not guarantee StorageVersions will be updated when the CRD gets updated.

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -919,6 +928,51 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
+		// This means stale records can still exist in the cluster if an apiserver
+		// hasn't got any request. All apiservers have to at least serve one
+		// request (can be a GET) for a CR's storage version record to be
+		// up-to-date, and to trigger storage migration.
```
An alternative to the owner-reference approach is to let `crdHandler` garbage collect storageversion whenever we tear down old storage. That approach is more responsive, but it's more expensive (apiextensions-apiserver replicas conflict with each other in HA cluster) and only helps with some rare corner cases.

roycaihw replied:
Specifically, in an HA cluster, after a CRD gets encoding version update:

1. if all three server have served requests with new storage, storage migration will be triggered for both approaches. 
2. if only one or two of the three servers have served requests with new storage, storage migration won't happen for the owner-reference approach, but will for the alternative
3. if none of the three servers has served requests with new storage, storage migration won't happen for the owner-reference approach; it will for the alternative **by a chance**: 
  a. if the storage migrator observed the old StorageVersion object getting shrunk to one/two server records, it will trigger the migration
  b. if the watch events get compressed (e.g. informer reconnection) and the storage migrator only saw a DELETE event, it won't trigger the migration

3 (updating the schema without even read after the update) should be extremely rare. I don't expect 2 to happen assuming enough number (>3) of requests have made to the servers and the load balancer works reasonably. 

That said, we can let the storage migrator send read requests periodically to bump the chances of 1 & 2. Also the alternative can be a future improvement if one region being starved is a concern.

sttts replied:
Having this logic in the handler chain feels misplaced. Why is that? Just register in the event handlers for CRDs (where we also install the handler in the first place).

sttts replied:
Also above we have event handlers: https://github.com/kubernetes/kubernetes/pull/96403/files#diff-01e7a2059fe62e95ec32e2589c01b7218100de327647c641c334d38f2c5aa06eL206.

roycaihw replied:
not sure what is `pkg/controllers/storageversion`. 

let me think about using the event handlers to create/update storage versions. I don't want to use the event handlers to garbage collect storage versions because of reasons in https://github.com/kubernetes/kubernetes/pull/96403/files#r520386107.

roycaihw replied:
I agree using the event handlers to create/update storage versions works better, and solves https://github.com/kubernetes/kubernetes/pull/96403/files#r520403155. Will update the PR

neolit123 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:
```diff
@@ -188,6 +192,15 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
+		utilfeature.DefaultFeatureGate.Enabled(genericfeatures.APIServerIdentity) {
+		kubeclientset, err := kubernetes.NewForConfig(s.GenericAPIServer.LoopbackClientConfig)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create kubernetes clientset: %v", err)
```
looks like the bots picked me from /test

should Kubernetes be uppercase here, or can it be dropped similar to line 173:
> return nil, fmt.Errorf("failed to create clientset: %v", err)

roycaihw replied:
Thanks for reviewing! `kubernetes` is the package name: `k8s.io/client-go/kubernetes.NewForConfig`. I have it to distinguish from L173, which is the apiextensions clientset: `k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset.NewForConfig`. Changed the error to `failed to create clientset for storage versions` to make it less confusing.

neolit123 commented on test/integration/storageversion/OWNERS:
```diff
@@ -0,0 +1,2 @@
+approvers:
+- roycaihw
```
potentially the SIG label should be here:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apimachinery/OWNERS#L25-L26

not sure if there should be more approvers added here too. ideally, yes.

roycaihw replied:
added the label and @caesarxuchao in approvers. Sadly I couldn't find a sig-api-machinery-approvers alias

neolit123 commented on test/integration/storageversion/storage_version_test.go:
```diff
@@ -0,0 +1,114 @@
+	// Send a request to make the server create handler and update storage version
+	_, err = dynamicClient.Resource(gvr).Namespace("default").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("unexpected error when listing foos: %v", err)
```
replace "foos" with gvr.Resource or just "resources"?
IIRC, the error here should already include what failed to be listed.

neolit123 commented on test/integration/storageversion/storage_version_test.go:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	var storageVersion apiserverinternalv1alpha1.ServerStorageVersion
+	if err := wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
```
do you see a case where 10 seconds might not be enough - i.e. to not introduce a flake?

roycaihw replied:
good point. Actually we don't need the wait. I was doing something else and forgot about it. Removed

neolit123 commented on test/integration/storageversion/storage_version_test.go:
```diff
@@ -0,0 +1,114 @@
+		t.Fatalf("failed to get storage version for custom resources: %v", err)
+	}
+	if !strings.HasPrefix(storageVersion.APIServerID, "kube-apiserver-") {
+		t.Fatalf("apiserver ID doesn't contain kube-apiserver- prefix, has: %v", apiserverID)
```
i think apiserverID is "" at this point.

roycaihw replied:
It's populated by the apiserver-identity feature: https://github.com/kubernetes/kubernetes/blob/3146daf4334507f9bbc99b2ae333f9acb76fe3c3/staging/src/k8s.io/apiserver/pkg/server/config.go#L299

fedebongio commented:
/triage accepted

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:
```diff
@@ -188,6 +192,15 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 	}
 	establishingController := establish.NewEstablishingController(s.Informers.Apiextensions().V1().CustomResourceDefinitions(), crdClient.ApiextensionsV1())
+	var sc storageversion.Client
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.StorageVersionAPI) &&
```
nit: do we even need these feature gate checks? Wouldn't it be enough to do that inside the handler?

roycaihw replied:
had the feature gates merely because `kubernetes.NewForConfig` could in theory return an error, and I don't want the apiextensions apiserver to fail start unexpectedly

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -508,6 +520,9 @@ func (r *crdHandler) updateCustomResourceDefinition(oldObj, newObj interface{})
 
 	klog.V(4).Infof("Updating customresourcedefinition %s", newCRD.Name)
 	r.removeStorage_locked(newCRD.UID)
+	if err := r.updateStorageVersionFor(newCRD); err != nil {
```
Am a little worried to have this inside of the lock as it calls out via a client (and potentially blocks). Same above.

roycaihw replied:
discussed offline. Decoupled the update logic and the lock

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -468,7 +477,17 @@ func (r *crdHandler) createCustomResourceDefinition(obj interface{}) {
+			// make sure no in-fight request to old storage exists
+			<-tearDownFinished
+			if err := r.updateStorageVersionFor(crd); err != nil {
+				utilruntime.HandleError(err)
```
I'd like the update to block the handler chain on update failures, otherwise the storage migration won't be triggered when it should. Here I accept to not block because 1. the chance of storage version update failure is low 2. it's not ideal but crd authors can self-heal by sending another CRD-versioning update

sttts replied:
If we know that the storage version really changed, and if we had locks per CRD, we could be much more aggressive with blocking until this call returns.

roycaihw replied:
After some thoughts, I think we might not need a hard block for custom resources. For built-in resources, we require the apiserver to update storageversion before serving writes, because otherwise one apiserver can sneak in writing some data with a different storageversion, and die without updating storageversion. In that case the storage migrator won't be able to tell and won't trigger migration.

For custom resources, all apiserver replicas write the same storageversion update when a CRD changes its storageversion. It's unlikely for all apiservers to fail update storageversion (conflict won't cause that; restarted server will retry update; the only way is to kill all servers forever, which doesn't make sense).

> the storage migration won't be triggered when it should

It's totally fine for storage migration to kick in after the servers have served some write requests with new storage, as long as:
1. it can tell a migration is needed when encoding versions become consistent-- one apiserver succeeded to update storageversion and change the resource version is good enough
2. encoding versions eventually become consistent-- maybe we just need to allow more retries if the failure rate bothers us.

roycaihw replied:
@caesarxuchao convinced me that blocking new handler on storageversion update is necessary, and this PR doesn't have to get in 1.20 for StorageVersionAPI alpha. I will revisit locks per CRD later. Thanks for the review

sttts commented:
I am fine with the factoring. I leave final review to @caesarxuchao. Ping me when you need an approval.

roycaihw commented:
/retest
/test pull-kubernetes-e2e-gce-alpha-features

caesarxuchao commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -1338,3 +1374,48 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *goopenapispec.Swagger, crd *a
+	encodingVersion := crd.Spec.Group + "/" + storageVersion
+	var decodableVersions []string
+	for _, v := range crd.Spec.Versions {
+		if !v.Served {
```
The doc says "Served is a flag enabling/disabling this version from being served via REST APIs", even if a version is not served, it's still decodable, right?

sttts replied:
it is

caesarxuchao commented on test/integration/storageversion/storage_version_test.go:
```diff
@@ -0,0 +1,138 @@
+	}
+	if len(storageVersion.DecodableVersions) != 1 {
+		t.Errorf("unexpected number of decodable versions, expected 1 version, got: %v", storageVersion.DecodableVersions)
+	}
```
Can you also verify the decodableVersions[0]?

caesarxuchao commented on test/integration/storageversion/storage_version_test.go:
```diff
@@ -0,0 +1,138 @@
+		if len(storageVersion.DecodableVersions) != 2 {
+			lastErr = fmt.Errorf("unexpected number of decodable versions, expected 2 versions, got: %v", sv.Status.StorageVersions)
+			return false, nil
+		}
```
Can you verify the contents of DecodableVersions?

caesarxuchao commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -468,7 +477,17 @@ func (r *crdHandler) createCustomResourceDefinition(obj interface{}) {
+	tearDownFinished := r.removeStorage_locked(crd.UID)
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+		go func() {
```
Why is this running in a goroutine? Once the customStorageLock is unlocked, CR requests can go through. Then a CR encoded in a version different than the version recorded in the storageVersion API will be written to the etcd.

sttts replied:
This code is doing client calls. We have the lock in this moment for the whole CRD apiserver storage map. I.e. a blocked client call might block other requests of other CRDs. This go routine will decouple those.

roycaihw commented:
@sttts @caesarxuchao I updated the PR to block new CR handler on CRD storage version update. Please take a look

roycaihw commented:
/retest

caesarxuchao commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -468,7 +521,25 @@ func (r *crdHandler) createCustomResourceDefinition(obj interface{}) {
+		q <- &storageVersionUpdateRequest{
+			crd:  crd,
+			wait: tearDownFinished,
+		}
```
Why do we need to update the storage version here? I thought it's enough to do a storageVersionUpdate in `getOrCreateServingInfoFor`.

Is the purpose forcing a storage version update even if there is no write requests? Is that necessary?

caesarxuchao replied:
Note that even if we don't do storage version updates here, we still need to enqueue the req to block later storage version updates until the tearDownFinished is closed.

roycaihw replied:
> Is the purpose forcing a storage version update even if there is no write requests?

yes, to update the storage version even if there is no new requests

caesarxuchao replied:
I thought we agreed to only update storage version after there is a write request.

roycaihw replied:
I thought differently :) Do you member why we want to do that?

If we only update storage version after there is a new read/write request (note: read requests also trigger underlying storage creation), the correctness of storage version depends on:
1. active CR traffic exists
2. in an HA cluster, the traffic gets routed to all API servers

caesarxuchao commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -1338,3 +1460,122 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *goopenapispec.Swagger, crd *a
+// Closing the channel ensures that we don't leak goroutine for the given
+// UID.
+func (r *crdHandler) stopStorageVersionUpdateQueue_locked(uid types.UID) {
+	if queue, ok := r.storageVersionUpdateQueues[uid]; ok {
```
If we add cancel() to req, we can just cancel all the remaining reqs.

caesarxuchao commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -1338,3 +1460,122 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *goopenapispec.Swagger, crd *a
+			// the old storage version update request can race against the new
+			// queue, and causing the storage version to go back in time.
+			// TODO(roycaihw): find a way to cancel the req when we drain the
+			// old queue (e.g. use context)
```
I think we should add a cancel() function to the req, and call cancel() for all remaining reqs when stopStorageVersionUpdateQueue_locked is called, then we don't have this race.

caesarxuchao replied:
Another use of the cancel() function is that when a new req is enqueued, we can cancel the existing storage version updates (and fail the pending CR write requests). Note that we still need to wait for pending teardown to finish draining.

roycaihw replied:
addressed the TODOs

roycaihw replied:
> add a cancel() function to the req

note that the cancel() function shouldn't be attached to a specific req, since the race here is about not being able to grab a req. I made all the req in one queue share the same context, and attach the cancel() function to the queue. 

> we can cancel the existing storage version updates (and fail the pending CR write requests)

I'm not fond of failing CR requests when we don't have to. The requests should wait for the storage version update and then proceed. The only place we have to fail in-flight requests is when a CRD is deleted-- we have to tear down the queue to avoid leaking goroutine, and we have to do it in a quick manner because we don't know if the CRD will be re-created and trigger new storage version updates.

caesarxuchao reviewed (commented):
In summary, two major comments:
1. if we need to do storage version updates in crdInformer's add handler and update handler.
2. add a cancel() function to the storageVersionUpdateRequest, it can help solve the race you mentioned, also it help avoid unnecessary storage version updates when the CRD's storage version is flipped quickly.

caesarxuchao commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -1338,3 +1477,110 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *goopenapispec.Swagger, crd *a
+			}
+			if req.processed != nil {
+				select {
+				case <-ctx.Done():
```
Why not checking the ctx.Done() at the beginning? Then we can skip updating storage version if the CRD is being deleted.

roycaihw replied:
> skip updating storage version if the CRD is being deleted

client-go is responsible for doing that. 

If we check ctx.Done() at the beginning, there could be a case where the CRD is being deleted after the check but before `r.updateStorageVersionFor`. Still, client-go has to evaluate the context and skip updating storage version.

caesarxuchao replied:
Got it. Can you test if client-go actually respects this ctx?

roycaihw replied:
Yes, it does
```
--- a/staging/src/k8s.io/client-go/examples/out-of-cluster-client-configuration/main.go
+++ b/staging/src/k8s.io/client-go/examples/out-of-cluster-client-configuration/main.go
@@ -61,7 +61,9 @@ func main() {
                panic(err.Error())
        }
        for {
-               pods, err := clientset.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
+               ctx, cancel := context.WithCancel(context.Background())
+               cancel()
+               pods, err := clientset.CoreV1().Pods("").List(ctx, metav1.ListOptions{})
                if err != nil {
                        panic(err.Error())
                }

$ go run examples/out-of-cluster-client-configuration/main.go --kubeconfig=/var/run/kubernetes/admin.kubeconfig
panic: context canceled
```

caesarxuchao replied:
Thanks for the confirmation.

Adding a check of ctx.Done() at the beginning let us skip unnecessary wait on `<-req.wait`.

caesarxuchao commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -1338,3 +1477,110 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *goopenapispec.Swagger, crd *a
+		//      CRs remain, the new CRD will update new storage version
+		klog.V(4).Infof("Cancelling the storage version update queue for CRD with UID %v.", uid)
+		queue.cancel()
+		for len(queue.q) > 0 {
```
If you take my previous comment, "Why not checking the ctx.Done() at the beginning?", then you don't need to drain the queue here.

roycaihw replied:
See https://github.com/kubernetes/kubernetes/pull/96403#discussion_r553047824. IMO, even if we check the ctx.Done() at the beginning, draining the queue actively still has its value compared to relying on `ctx.Done()` checks passively.

caesarxuchao replied:
Can you elaborate on the value of doing the draining here?
Also, since there are two consumers of the queue, do you need a lock before checking `len(queue.q)`?

roycaihw replied:
Relying on `ctx.Done()` checks means the queued requests will be processed and failed after the currently being-processed request hit one of the `ctx.Done()` checks. On the other hand, draining the queue allows us to fail the queued requests and release the resources immediately.

> there are two consumers of the queue
Nice catch! I added an unblocking select to handle the race.

caesarxuchao replied:
> Relying on ctx.Done() checks means the queued requests will be processed and failed after the currently being-processed request hit one of the ctx.Done() checks.

That's right, but I don't think it's much difference in practice, especially if you add a check of `ctx.Done()` at the beginning of the other consumer code. I would sacrifice a little performance for easier concurrency model.

caesarxuchao commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -940,6 +1052,33 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
+		// a too-frequently-updated CRD can block the entire API server. Here we
+		// give up updating storage version for this CRD and print a log.
+		// TODO(roycaihw): use warning instead of info when StorageVersionAPI graduates to beta/GA
+		klog.V(4).Infof("Skipping the storage version update for CRD with UID %v due to the queue being full (queue size: %v).",
```
This should be a V(2). This is expected to be rarely triggered, and when it's triggered, we do want to see it in the logs.

roycaihw replied:
Done.

caesarxuchao commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -1338,3 +1477,110 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *goopenapispec.Swagger, crd *a
+				close(req.abandoned)
+			}
+		}
+		close(queue.q)
```
Is it possible that a request is added to the queue after you checked `len(queue.q)==0` but before you close the queue?

It seems to be impossible because here we hold the customStorageLock and all enqueue operations are protected by the same lock. Can you double check and add a comment explaining why it's safe?

roycaihw replied:
It's impossible. Added a comment

roycaihw commented:
/retest

caesarxuchao commented on test/integration/storageversion/storage_version_test.go:
```diff
@@ -0,0 +1,147 @@
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+func TestStorageVersionCustomResource(t *testing.T) {
```
Have you verified that the apiextension-apiserver has the necessary RBAC permissions, e.g., to create/update storageVersion?

This occurred to me because many times we missed the RBAC checks because the integration test doesn't do that.

roycaihw replied:
yes, it uses the generic server loopback client

caesarxuchao commented on test/integration/storageversion/storage_version_test.go:
```diff
@@ -0,0 +1,147 @@
+	if err := kubeclient.InternalV1alpha1().StorageVersions().Delete(context.TODO(), gr, metav1.DeleteOptions{}); err != nil {
+		t.Errorf("failed to delete storage version: %v", err)
+	}
+}
```
Can you add a stress test? something like
```golang
for i:=0; i<10; i++ {
   // flip the storage version of the CRD;
   // create a few CRs;
}
// create a different CRD, verifies that the CRD creation is not blocked by storage version update of the first CRD;
// verify that the storage version of the first CRD eventually stablize to the expected version.
```

roycaihw replied:
done

caesarxuchao commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -360,6 +396,43 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+			return
+		// Unblock the requests if the storage version update takes a long time, otherwise
+		// CR requests may stack up and overwhelm the API server. We will benchmark the
+		// storage version update latency to adjust the timeout in future.
```
nit: in the future

Can you add a `TODO(roycaihw)` or link it to an issue?

caesarxuchao commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -360,6 +396,43 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+		// NOTE: currently the graceful CRD deletion waits 1s for in-flight requests
+		// to register themselves to the wait group. Ideally the storage version update should
+		// not cause the requests to miss the 1s window; otherwise the requests may
+		// fail ungracefully (e.g. it may happen if the CRD was deleted immediately after the first CR request
```
nit: can you break the line at the same length?

caesarxuchao commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -360,6 +396,43 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+		// to register themselves to the wait group. Ideally the storage version update should
+		// not cause the requests to miss the 1s window; otherwise the requests may
+		// fail ungracefully (e.g. it may happen if the CRD was deleted immediately after the first CR request
+		// establishes the underlying storage).
```
nit: Can you move this "NOTE" to the beginning of the `select`? Given that `storageVersionUpdateTimeout` is 15s now, this timeout case won't help guarantee the graceful shutdown anyway.

caesarxuchao reviewed (commented):
Thanks, @roycaihw! I added some nits and requested for a stress test.

This mostly LGTM. Let's involve @sttts in the review process now.

sttts commented on staging/src/k8s.io/apiserver/pkg/storageversion/updater.go:
```diff
@@ -121,13 +124,19 @@ func setStatusCondition(conditions *[]v1alpha1.StorageVersionCondition, newCondi
-// updateStorageVersionFor updates the storage version object for the resource.
-func updateStorageVersionFor(c Client, apiserverID string, gr schema.GroupResource, encodingVersion string, decodableVersions []string) error {
+// UpdateStorageVersionFor updates the storage version object for the resource.
+func UpdateStorageVersionFor(ctx context.Context, c Client, apiserverID string, gr schema.GroupResource, encodingVersion string, decodableVersions []string, postProcessFunc processStorageVersionFunc) error {
```
doc what postProcessFunc is and when it is called. I assume it is called before updating the status of the CRD.

roycaihw replied:
it modifies the StorageVersion object before the object gets created/updated. It allows customizing the object (e.g. setting an owner reference pointing to a CRD) as a last step

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -89,6 +92,11 @@ import (
 
+const (
+	maxStorageVersionUpdateRequests = 10
+	storageVersionUpdateTimeout     = 15 * time.Second
```
doc what this timeout means?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -89,6 +92,11 @@ import (
 )
 
+const (
+	maxStorageVersionUpdateRequests = 10
```
doc what these 10 are? Max retries in case of what? Conflicts? General errors?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -134,6 +142,25 @@ type crdHandler struct {
+
+type storageVersionUpdateQueue struct {
+	cancel context.CancelFunc
+	q      chan *storageVersionUpdateRequest
```
can this be a one-way channel? Would make ownership clear.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -134,6 +142,25 @@ type crdHandler struct {
+}
+
+type storageVersionUpdateQueue struct {
+	cancel context.CancelFunc
```
would expect this on a request, not on a queue. What does cancelling mean here?

roycaihw replied:
all the requests in one queue sharing the same context. Cancelling means aborting all the pending requests in the queue. 

It's used when a CRD is deleted: we close the queue for that CRD, to make sure the associated goroutine stops; we also don't know if the CRD will be re-created, bring up a new queue + goroutine, and race new requests with the old requests-- therefore we abort all the pending requests to avoid such race.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -134,6 +142,25 @@ type crdHandler struct {
+	apiserverID          string
+	storageVersionClient storageversion.Client
+
+	storageVersionUpdateQueues map[types.UID]*storageVersionUpdateQueue
```
doc what these queues do? Who sends anything? Who receives it?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -134,6 +142,25 @@ type crdHandler struct {
+	storageVersionUpdateQueues map[types.UID]*storageVersionUpdateQueue
+}
+
+type storageVersionUpdateRequest struct {
```
godoc

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -134,6 +142,25 @@ type crdHandler struct {
+	abandoned chan<- struct{}
+}
+
+type storageVersionUpdateQueue struct {
```
godoc

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -360,6 +396,43 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		)
 		return
 	}
+	// Make sure storage version is updated before we serve write requests
```
empty line above

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -360,6 +396,43 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+			})
+		}
+	}
 	switch {
```
empty line above

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -165,6 +192,10 @@ type crdInfo struct {
 	waitGroup *utilwaitgroup.SafeWaitGroup
+
+	storageVersionUpdateProcessed      <-chan struct{}
+	storageVersionUpdateAbandoned      <-chan struct{}
```
doc these. What do they mean? Who closes them?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -360,6 +396,43 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+			utilruntime.HandleError(fmt.Errorf("failed to update storage version for custom resource, group: %v, resource: %v",
+				requestInfo.APIGroup, requestInfo.Resource))
+			// Storage version update failed because the CRD was deleted. Here
+			// we return a 503 to fail this in-flight request and let the client
```
non-existing resources return 403, don't they?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -360,6 +396,43 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+		// Unblock the requests if the storage version update takes a long time, otherwise
+		// CR requests may stack up and overwhelm the API server.
+		// TODO(roycaihw): benchmark the storage version update latency to adjust the timeout.
+		case <-time.After(storageVersionUpdateTimeout):
```
shouldn't we skip the whole waiting logic (the whole select) after this is hit once? I would rather prefer a fixed timestamp in the crdInfo and wait for the difference from now to that timestamp. I.e. some duration that has the same starting point, and not restarting on every requests as it is done here.

sttts replied:
Something like `<-time.After(crdInfo.storageVersionUpdateTimeout.Sub(time.Now()))`

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -468,7 +541,25 @@ func (r *crdHandler) createCustomResourceDefinition(obj interface{}) {
 	defer r.customStorageLock.Unlock()
 	// this could happen if the create event is merged from create-update events
-	r.removeStorage_locked(crd.UID)
+	tearDownFinished := r.removeStorage_locked(crd.UID)
```
usually channels have a `Ch` postfix.

sttts commented:
Did a first pass of this. I am not convinced about the factoring. It adds complicated synchronization logic to an already complex CRD handler. In the current shape I have strong concerns to merge this as is, not even as alpha.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -1338,3 +1500,116 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *goopenapispec.Swagger, crd *a
+
+	queue := make(chan *storageVersionUpdateRequest, maxStorageVersionUpdateRequests)
+	ctx, cancel := context.WithCancel(context.TODO())
+	r.storageVersionUpdateQueues[uid] = &storageVersionUpdateQueue{
```
could storageVersionUpdateQueues be something completely independent from crdHandler ? If this go routine management was done in a self-contained unit with a clear interface, the code would become considerably simpler.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -360,6 +396,43 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+					crdInfo.spec.Group, crdInfo.spec.Names.Plural)
+			})
+		}
+	}
```
wondering whether one could move this whole block of code into `getOrCreateServingInfoFor`, maybe as `getOrCreateServingInfoForWrite`.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -494,11 +585,29 @@ func (r *crdHandler) updateCustomResourceDefinition(oldObj, newObj interface{})
+			utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+			q := r.getOrCreateStorageVersionUpdateQueue_locked(newCRD.UID)
+			if len(q) == maxStorageVersionUpdateRequests {
+				// When the channel buffer is full, writing to the channel becomes blocking.
```
when can this happen? On storage version changes should be enough, shouldn't it? Now it's on every update of the CRD?

roycaihw replied:
update happens at:

1. new storage creation
2. CRD update event: old storage doesn't exist
3. CRD update event: old storage exists, and the storage version changes
4. CRD create event: old storage doesn't exist
5. CRD create event: old storage exists, and the storage version changes

2, 3, 4, 5 exist only to make sure if there is no CR request, or if CR requests aren't hitting all servers in an HA cluster, the StorageVersion record for each server still gets updated.

we don't send update requests when:
1. CRD update event: old storage with same storage version exists
2. CRD create event: old storage with same storage version exists

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -601,6 +672,46 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
+		case <-ret.storageVersionUpdateProcessedCh:
+		case <-ret.storageVersionUpdateAbortedCh:
+			// Storage version update failed because the CRD was deleted. Here
+			// we return a 404 to fail this in-flight request.
```
I believe this must be a 403, not a 404. Please double check. Afaik for the garbage collector it makes a difference: 404 = gone. 403 = unknown owner reference. @deads2k can comment.

roycaihw replied:
> 403 StatusForbidden
Indicates that the server can be reached and understood the request, but refuses to take any further action, because it is configured to deny access for some reason to the requested resource by the client.
Suggested client recovery behavior: Do not retry. Fix the request.

I agree 404 is better than 403. 404 allows us to provide more detail in the error message about why the request got rejected. Plus the server truly did understand the request. I changed to 403

I'm not so sure about the recovery behavior though. There isn't much the CR client can do to fix the request (assuming a different user deleted the CRD). A 503 can let the CR client to simply retry, and it will get either a 404 (CRD teardown finished) or a success (CRD re-created) later.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -601,6 +672,46 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
+		// Unblock the requests if the storage version update takes a long time, otherwise
+		// CR requests may stack up and overwhelm the API server.
+		// TODO(roycaihw): benchmark the storage version update latency to adjust the timeout.
+		case <-time.After(ret.storageVersionUpdateTimeout.Sub(time.Now())):
```
time.After hopefully works with negative durations 🤞

roycaihw replied:
yes, zero and negative durations are ignored

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -504,11 +555,19 @@ func (r *crdHandler) updateCustomResourceDefinition(oldObj, newObj interface{})
 	if !found {
+		if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+			utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+			r.storageVersionManager.UpdateStorageVersionFor(newCRD, nil, nil, nil)
```
maybe I missed it, do we check anywhere that the update is necessary? Would feel better to have only very specific cases where we queue an update.

sttts replied:
Line 515 below is new, right? That would cause the skip because the code above only fires when `!found` is true and this is never the case on normal updates? Maybe this is worth a oneliner comment here.

roycaihw replied:
this is case 2), and line 515 is case 4) in https://github.com/kubernetes/kubernetes/pull/96403#discussion_r558146492. This happens when a CRD gets updated before any CR request hits the server. 

re-organized the code and added comments: we skip storage version update if an old-and-exact-same storage exists, in which case we know we have done the same storage version update already; otherwise we do a storage version update with the latest info in the watch event.

sttts replied:
:+1:

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -601,6 +672,46 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
+
+	// Double check the feature gate is enabled
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) &&
```
nit: consider early return. Lets you unindent all the code below

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,254 @@
+// customStorageLock, to make sure the storage version update is in sync with
+// the CRD underlying storage change.
+type Manager interface {
+	// UpdateStorageVersionFor updates StorageVersion for the given CRD.
```
better: .... queues an update and returns immediately.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,254 @@
+// same CRD happen in the same order as the methods get invoked.
+// NOTE: the caller should hold a lock while calling the methods, supposedly the
+// customStorageLock, to make sure the storage version update is in sync with
+// the CRD underlying storage change.
```
this is not completely clear here, especially because `UpdateStorageVersionFor` just queues and returns.

roycaihw replied:
moved lock-related comments to crdHandler. The manager doesn't care about the lock

roycaihw commented:
/retest

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -950,6 +1086,17 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
+		!utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+		return ret, nil
+	}
+
```
// spawn storage version update in background and use channels to make handlers wait

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -165,6 +184,23 @@ type crdInfo struct {
 	storageVersion string
 
 	waitGroup *utilwaitgroup.SafeWaitGroup
+
```
```
storageVersionUpdate struct {
   ...
}
```

makes code a little bit prettier everywhere.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -601,6 +690,53 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
+	case <-ret.storageVersionUpdateAbortedCh:
+		// Storage version update failed because the CRD was deleted. Here
+		// we return a 403 to fail this in-flight request.
+		return nil, apierrors.NewForbidden(schema.GroupResource{
```
the forbidden error is misplaced here. Would rather return `nil, nil, true` here, and react to `abort==true` in the handler. The handler knows which error to return to the caller (403).

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -601,6 +690,53 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
+// indicating that the server forbids the CR request due to the CRD deletion.
+// Otherwise, any non-nil error returned by getOrCreateServingInfoFor may be
+// wrapped by the server as an internal error (500) before surfacing to the client.
+func (r *crdHandler) getOrCreateServingInfoForWrite(uid types.UID, name string) (ret *crdInfo, err error, aborted bool) {
```
by convention (err last) make it `*crdInfo, bool, error`.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,256 @@
+		klog.V(2).Infof("Skipping the storage version update for CRD with UID %v due to the queue being full (queue size: %v).",
+			crd.UID, updateQueueBufferSize)
+		if processedCh != nil {
+			close(processedCh)
```
does this case mean that another CRD state in time is queued already and will be used to update the storage version, but we still claim that it got processed? How do we guarantee consistency? Would rather expect to put just one update per UID into the queue, and replace the unprocessed update with a more up-to-date one, or even merge them.

roycaihw replied:
> one update per UID into the queue

In theory it's possible that a CRD's storage version changes from v1 to v2, and changes back to v1, while serving CR write requests in each stage. We cannot just update the storage version to v1 and unblock all the write requests, because the storage migrator won't be able to tell that v2 data exists in etcd. Therefore, there can be more than one update for a UID in the queue at some point.

roycaihw replied:
A possible optimization is to merge consecutive updates triggered by watch event handlers. I added a TODO

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,256 @@
+		q:      queue,
+		cancel: cancel,
+	}
+	go func() {
```
this need panic handling. Otherwise updates for a UID can be blocker forever (until process restart).

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,256 @@
+				case <-ctx.Done():
+					// The queue was cancelled. Potentially we didn't finish this
+					// storage version update.
+					close(update.abortedCh)
```
this panics on nil channel

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,256 @@
+// one and a new goroutine if necessary. The goroutine keeps processing updates
+// until the channel is closed.
+func (m *manager) getOrCreateUpdateQueue(uid types.UID) chan<- *update {
+	if queue, ok := m.updateQueues[uid]; ok {
```
don't we need any locking?

roycaihw replied:
the caller (crd handler) had a lock. Added a separate lock to the manager for better isolation.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,256 @@
+// TeardownFor closes the channel for the given UID. It ensures that we don't
+// leak goroutines.
+func (m *manager) TeardownFor(uid types.UID) {
+	if queue, ok := m.updateQueues[uid]; ok {
```
no locking?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -315,7 +353,27 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+		var aborted bool
+		crdInfo, err, aborted = r.getOrCreateServingInfoForWrite(crd.UID, crd.Name)
+		// The server aborted storage version update due to the CRD being deleted.
+		// The CR request is forbidden.
```
```
// Requests to non-existent resources return 403
```

Are we sure the only case for `abort==true` is CRD deletion?

roycaihw replied:
(other than the new "should not happen" panic handling) yes

liggitt replied:
* 403 doesn't seem right
* if the CRD is being deleted, but is pending cleanup of CRs, won't erroring here deadlock?

roycaihw replied:
> won't erroring here deadlock?

we don't block delete requests, and I'm not sure if the storage DestroyFunc goes through this code path.

roycaihw replied:
we won't deadlock because the CRD finalizer doesn't go through this code path to cleanup CRs

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go:
```diff
@@ -0,0 +1,256 @@
+	}
+}
+
+func (m *manager) updateStorageVersionFor(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) error {
```
we have 3 kinds of "UpdateStorageVersionFor" funcs. Maybe rename them to be more clear?

roycaihw committed:
API server updates storage version for custom resources

roycaihw committed:
integration test; add OWNERS file

roycaihw committed:
generated

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -950,6 +1100,20 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
+	processedCh := make(chan struct{})
+	abortedCh := make(chan struct{})
+	r.storageVersionManager.EnqueueStorageVersionUpdate(crd, nil, processedCh, abortedCh)
+	ret.storageVersionUpdate = &storageVersionUpdate{
```
here is a race. The `r.customStorage.Store(storageMap2)` must come laster *AFTER* `storageMap2` and `ret` are in its final state.

roycaihw replied:
nice catch! Fixed

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -601,6 +704,53 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
+	// Unblock the requests if the storage version update takes a long time, otherwise
+	// CR requests may stack up and overwhelm the API server.
+	// TODO(roycaihw): benchmark the storage version update latency to adjust the timeout.
+	case <-time.After(ret.storageVersionUpdate.timeout.Sub(time.Now())):
```
With this timeout we risk an inconsistent storage version state, right? Am still unsure whether is worse: bad storage version object or failing CRD writes.

@liggitt would like to hear your opinion.

liggitt replied:
I'm finding the handling of a CR really hard to reason about in this PR, especially with a long (O(seconds)) delay between constructing the serving info at the top of this method and returning to unblock serving

roycaihw replied:
> bad storage version object or failing CRD writes

@sttts I considered failing CRD writes a regression, but I'm okay with that direction if we think it's necessary. Let me add some change in that direction

sttts replied:
If we cannot write the StorageVersion object, we have big problems anyway. Do we care about CRD writes in this case?

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -315,7 +363,31 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	var crdInfo *crdInfo
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) &&
+		(requestInfo.Verb == "create" ||
+			requestInfo.Verb == "update" ||
+			requestInfo.Verb == "patch") {
```
* this checking here seems fragile... should we not intercept in the actual serveResource function?
* delete and deletecollection can also write
* the featuregates seem like they could be checked once at initialization time and compacted into a local boolean checked everywhere in this file instead of the double feature gate check

roycaihw replied:
> delete and deletecollection can also write

the goal of blocking writes before StorageVersion update is to make sure we don't accidentally persist data in some version that the storage migrator is unaware of. Deleted CRs are not persisted, so I think it's okay to unblock these requests.

> the featuregates seem like they could be checked once

done

liggitt replied:
> Deleted CRs are not persisted, so I think it's okay to unblock these requests.

They are if they have finalizers on them (a delete request populates metadata.deletionTimestamp and re-persists)

roycaihw replied:
> a delete request populates metadata.deletionTimestamp and re-persists

Ack. I will double check and think about the deadlock

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -601,6 +704,53 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
+	select {
+	case <-ret.storageVersionUpdate.processedCh:
+	case <-ret.storageVersionUpdate.abortedCh:
+		// Storage version update aborted. The caller knows which error
+		// to return to the client (403).
```
* how does a server/user get out of this error state?
* 403 does not seem like the correct error code... would "unavailable" or something similar be better?

roycaihw replied:
> how does a server/user get out of this error state

the error state is transient and only for in-flight requests. A CR client can retry: 
- if the CRD is gone forever, the client will get a 404. 
- if the CRD gets re-created, things continue to work

> would "unavailable" or something similar be better

had 503 originally (ref https://github.com/kubernetes/kubernetes/pull/96403#discussion_r559328435). I'm okay with 503. If we assume a CR client and a CRD client are the same user, then 403 (let the user fix the request/situation) also makes sense

roycaihw committed:
fix a race: make sure r.customStorage.Store comes after ret is final

roycaihw committed:
replace double feature gate check with non-nil interface check

roycaihw committed:
generated

liggitt commented:
I was hoping for a change more like this, where we intercept every write operation:

```diff
diff --git a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go
index b317e30ebd2..263b4524154 100644
--- a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go
+++ b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go
@@ -402,16 +402,31 @@ func (r *crdHandler) serveResource(w http.ResponseWriter, req *http.Request, req
 			responsewriters.ErrorNegotiated(err, Codecs, schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}, w, req)
 			return nil
 		}
+		if err := crdInfo.WaitForConsistentStorageVersion(req.Context()); err != nil {
+			return err
+		}
 		return handlers.CreateResource(storage, requestScope, r.admission)
 	case "update":
+		if err := crdInfo.WaitForConsistentStorageVersion(req.Context()); err != nil {
+			return err
+		}
 		return handlers.UpdateResource(storage, requestScope, r.admission)
 	case "patch":
+		if err := crdInfo.WaitForConsistentStorageVersion(req.Context()); err != nil {
+			return err
+		}
 		return handlers.PatchResource(storage, requestScope, r.admission, supportedTypes)
 	case "delete":
 		allowsOptions := true
+		if err := crdInfo.WaitForConsistentStorageVersion(req.Context()); err != nil {
+			return err
+		}
 		return handlers.DeleteResource(storage, allowsOptions, requestScope, r.admission)
 	case "deletecollection":
 		checkBody := true
+		if err := crdInfo.WaitForConsistentStorageVersion(req.Context()); err != nil {
+			return err
+		}
 		return handlers.DeleteCollection(storage, checkBody, requestScope, r.admission)
 	default:
 		responsewriters.ErrorNegotiated(
@@ -430,8 +445,14 @@ func (r *crdHandler) serveStatus(w http.ResponseWriter, req *http.Request, reque
 	case "get":
 		return handlers.GetResource(storage, requestScope)
 	case "update":
+		if err := crdInfo.WaitForConsistentStorageVersion(req.Context()); err != nil {
+			return err
+		}
 		return handlers.UpdateResource(storage, requestScope, r.admission)
 	case "patch":
+		if err := crdInfo.WaitForConsistentStorageVersion(req.Context()); err != nil {
+			return err
+		}
 		return handlers.PatchResource(storage, requestScope, r.admission, supportedTypes)
 	default:
 		responsewriters.ErrorNegotiated(
@@ -450,8 +471,14 @@ func (r *crdHandler) serveScale(w http.ResponseWriter, req *http.Request, reques
 	case "get":
 		return handlers.GetResource(storage, requestScope)
 	case "update":
+		if err := crdInfo.WaitForConsistentStorageVersion(req.Context()); err != nil {
+			return err
+		}
 		return handlers.UpdateResource(storage, requestScope, r.admission)
 	case "patch":
+		if err := crdInfo.WaitForConsistentStorageVersion(req.Context()); err != nil {
+			return err
+		}
 		return handlers.PatchResource(storage, requestScope, r.admission, supportedTypes)
 	default:
 		responsewriters.ErrorNegotiated(
```

that contains the intercepting logic to a single method and leaves the rest of the custom resource handling logic alone

in that single method, we can:
* no-op if the feature is disabled or if there's only a single storageVersion (e.g. nothing to converge on)
* return fast once servers converge
* wait, honoring request timeout (or optionally a failsafe)

deads2k commented:
As I recall the KEP, our goal for CRs is that we don't ever storage a CR with an encoding version that doesn't match what is stored in the API.  This is slightly more complicated than the built-in resources since the encoding version can change without restarting.  However, I think we could simplify slightly with a flow like

1. create storage for CR as normal
2. read-only requests are allowed immediately
3. first mutating request for a CR arrives
4. create or update the storage version.  If this fails, fail the request
5. accept mutating requests

I don't know exactly where the code change would be offhand, but I don't know that we need asynchronous writing in this case.  The first few mutating requests after an encoding version change would have higher than normal latency, but there are not that many of them.

As I recall, there's a controller-y aspect to writing for the built-in resources.  I think a similar thing could be built using a cache that checks each request to see if it matches and does the synchronous write then.

roycaihw commented:
@deads2k There could be some race conditions. Imagine: 

3. ...
4. updating the storage version to v1. 
5. CRD got updated to v2 
6. apiserver saw the watch event, removed the v1 storage from the storage map, and put it to graceful teardown
7. (the v1 storage could still update the storage version to v1, and finish serving the in-flight mutating requests)
8. new mutating request came in
9. apiserver created v2 storage, updating the storage version to v2

The v2 update might happen before the v1 update. In that case the storage version ends up being v1. 

Or assuming the storage version updates happened in the right order (first v1, then v2), but some v1 in-flight mutating requests got served after the v2 storage version update. Furthermore if a storage migration happened after the v2 update and before the v1 in-flight requests, v1 data would get persisted in etcd. 

The asynchronous channel makes sure the updates happen in the right order, and latter updates begin only after the former ones' graceful teardown finishes.

roycaihw commented:
> I was hoping for a change more like this, where we intercept every write operation:

@liggitt Thanks. That's doable. I will update the PR. We also need same logic in `serveScale` and `serveStatus`.

roycaihw committed:
unify processedCh and abortedCh

When a storage version update is aborted because the CRD was deleted, we
don't need to reject in-flight CR requests. We can safely allow these
requests to proceed, because the CRD finalizer guarantees no write
requests can succeed after a CRD is deleted-- CREATE requests are 405
NOT ALLOWED, other requests will get 404 NOT FOUND.

roycaihw committed:
move wait logic to serveResource, intercept every write operation

roycaihw committed:
make sure the chance for an existing race is low

In theory, "CRD deletion and CRD finalizer tearing-down CRs" could happen
in between 1. crdHandler saw CRD terminating=false and 2. crdHandler
serves CR create request.

This commit makes crdHandler read the latest CRD from the cache (shared
with CRD finalizer) right before serving the create request, to make
sure the gap is as narrow as possible, and the chance for the race is
still low.

roycaihw committed:
make CRD storage version update always retry

roycaihw commented:
I checked CRD deletion workflow-- we don't need to reject CR writes if a CRD is deleted, see commit message https://github.com/kubernetes/kubernetes/pull/96403/commits/8661f73677c685c770102f59dc7c0ea96098a786 for more detail. One concern raised was https://github.com/kubernetes/kubernetes/issues/99181. I had a mitigation in https://github.com/kubernetes/kubernetes/pull/96403/commits/6ddc1eb1fb3c85a241a87ccf66c142be578ff4e3.

liggitt commented:
> the CRD finalizer guarantees no write requests can succeed after a CRD is deleted

are you sure? updates have to be allowed in order to remove finalizers on CR instances blocking cleanup, right?

roycaihw commented:
> updates have to be allowed in order to remove finalizers on CR instances blocking cleanup, right?

The cleanup goes directly to the [storage](https://github.com/kubernetes/kubernetes/blob/033d950152dc83fcd9afb546f8c112feb9e2c6f7/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go#L218), so it won't be blocked by the http handler. 

> by "after a CRD is deleted"

I mean actually deleted. This is when crdHandler gets a DELETE watch event, starts tearing down the storage. The CRD finalizer controller did its job already, so we can unblock write requests and know they won't succeed.

roycaihw commented:
> updates have to be allowed in order to remove finalizers on CR instances blocking cleanup, right?

> The cleanup goes directly to the storage, so it won't be blocked by the http handler.

I may be wrong. Let me double check. 

But even if CR cleanup needs to go through the http handler, we won't deadlock. The order would be:
1. storage version update succeeds
2. CR cleanup finishes
3. CRD is actually deleted
4. `crdHandler` sees the DELETE event, aborts storage version update (no-op since the update already succeeded), unblocks write requests (no-op)
